tutorial/source/tensor_shapes.ipynb

```diff
@@ -0,0 +1,555 @@
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "d = MultivariateNormal(torch.zeros(3), torch.eye(3, 3))\n",
+    "assert d.batch_shape == ()\n",
+    "assert d.event_shape == (3,)\n",
+    "x = d.sample()\n",
+    "assert x.shape == (3,)            # == batch_shape + event_shape\n",
+    "assert d.log_prob(x).shape == ()  # == batch_shape"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Reshaping distributions <a class=\"anchor\" id=\"dist-reshape\"></a>\n",
+    "\n",
+    "In Pyro you can treat a univariate distribution as multivariate by calling the `.reshape(extra_event_dims=_)` property."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "d = Bernoulli(0.5 * torch.ones(3,4)).reshape(extra_event_dims=1)\n",
+    "assert d.batch_shape == (3,)\n",
+    "assert d.event_shape == (4,)\n",
+    "x = d.sample()\n",
+    "assert x.shape == (3, 4)\n",
+    "assert d.log_prob(x).shape == (3,)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "While you work with Pyro programs, keep in mind that samples have shape `batch_shape + event_shape`, whereas `.log_prob(x)` values have shape `batch_shape`. You'll need to ensure that `batch_shape` is carefully controlled by either trimming it down with `.reshape(extra_event_dims=n)` or by declaring dimensions as independent via `pyro.iarange`.\n",
+    "\n",
+    "### It is always safe to assume dependence <a class=\"anchor\" id=\"assume-dependence\"></a>\n",
+    "\n",
+    "Often in Pyro we'll declare some dimensions as dependent even though they are in fact independent, e.g.\n",
+    "```py\n",
+    "pyro.sample(\"x\", dist.Normal(0, 1).reshape([10], extra_event_dims=1))\n",
+    "```\n",
+    "This is useful for two reasons: First it allows us to easily swap in a `MultivariateNormal` distribution later. Second it simplifies the code a bit since we don't need an `iarange` (see below) as in\n",
+    "```py\n",
+    "with pyro.iarange(\"x_iarange\", 10):\n",
+    "    pyro.sample(\"x\", dist.Normal(0, 1).reshape([10]))\n",
+    "```\n",
+    "The difference between these two versions is that the second version with `iarange` informs Pyro that it can make use of independence information when estimating gradients, whereas in the first version Pyro must assume they are dependent (even though the normals are in fact independent). This is analogous to d-separation in graphical models: it is always safe to add edges and assume variables *may* be dependent (i.e. to widen the model class), but it is unsafe to assume independence when variables are actually dependent (i.e. narrowing the model class so the true model lies outside of the class, as in mean field). In practice Pyro uses reparameterized gradient estimators for `Normal` distributions so both gradient estimators have the same performance. "
```

Comment:
" In practice Pyro uses reparameterized gradient estimators..."

+ ... when using SVI ... or the like